src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,254 @@
+interrupt execution the way Unix signals do):
+
+```rust
+static mut PRINT_WAKER: Option<Waker> = None;
+
+extern fn callback() {
+    if let Some(waker) = unsafe { PRINT_WAKER.as_ref() } {
+        waker.wake_by_ref();
+    }
+}
+```
+
+She then modifies `poll` to set `PRINT_WAKER`:
+
+```rust
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        if kernel_is_print_done() {
+            return Poll::Ready(kernel_get_buffer_back());
+        }
+        unsafe { PRINT_WAKER = Some(cx.waker()); }
+        Poll::Pending
+    }
+```
+
+`PRINT_WAKER` is stored in `.bss`, which occupies space in RAM but not flash. It
+is two words in size. It points to a
+[`RawWakerVTable`](https://doc.rust-lang.org/stable/core/task/struct.RawWakerVTable.html)
+that is provided by the executor. `RawWakerVTable` does not appear to be
+designed for a no-`alloc` environment. In a no-`alloc` environment, `drop` and
+`clone` are generally no-ops, and `wake`/`wake_by_ref` seem like duplicates.
+Looking at `RawWakerVTable` made Barbara think about the size impact of using
+`Future`.
+
+Barbara decided the overhead would be worth it if it made it easy to work with
+future combinators from external crates *and* could function as a lightweight
+form of multithreading. To support this, Barbara decides to implement an
+executor designed to function like a background thread. Because `alloc` is not
+available, this executor contains a single future. The executor has a `spawn`
+function that accepts a future and starts running that future (overwriting the
+existing future in the executor if one is already present). Once started, the
+executor runs entirely in kernel callbacks.
+
+Barbara identifies several factors that add branching and error handling code to
+the executor:
+
+1. `spawn` should be a safe function, because it is called by high-level
+   application code. However, that means it can be called by the future it
+   contains. If handled naively, this would result in dropping the future while
+   it executes. Barbara adds runtime checks to identify this situation.
+1. `Waker` is `Sync`, so on a multithreaded system, a future could give another
+   thread access to its `Waker` and the other thread could wake it up. This
+   could happen while the `poll` is executing, before `poll` returns
+   `Poll::Pending`. Therefore, Barbara concludes that if `wake` is called while
+   a future is being polled then the future should be re-polled, even if the
+   current `poll` returns `Poll::Pending`. This requires putting a retry loop
+   into the executor.
+1. A kernel callback may call `Waker::wake` after its future returns
+   `Poll::Ready`. After `poll` returns `Poll::Ready`, the executor should not
+   `poll` the future again, so Barbara adds code to ignore those wakeups. This
+   duplicates the "ignore spurious wakeups" functionality that exists in the
+   future itself.
+
+Ultimately, this made the executor logic [quite
```

Comment:
I don't think less than 100 LoC with lots of comments is "quite nontrivial". Certainly concurrent code is never easy. But for an executor in general this is fairly simple.

Reply:
I rephrased this to be more accurate. What I meant by "quite nontrivial" is it generated 96 bytes of code.

Reply:
This was a good change. I would have interpreted "quite nontrivial" differently as well.